modules/clickhouse/clickhouse_test.go

```diff
@@ -278,75 +279,69 @@ func TestClickHouseWithZookeeper(t *testing.T) {
 	defer conn.Close()
 
 	// perform assertions
-	data, err := performReplicatedCRUD(conn)
+	data, err := performReplicatedCRUD(t, conn)
 	require.NoError(t, err)
 	assert.Len(t, data, 1)
 }
 
-func performReplicatedCRUD(conn driver.Conn) ([]Test, error) {
-	var (
-		err error
-		res []Test
-	)
-
-	err = backoff.Retry(func() error {
-		err = conn.Exec(context.Background(), "CREATE TABLE replicated_test_table (id UInt64) ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/mdb.data_transfer_cp_cdc', '{replica}') PRIMARY KEY (id) ORDER BY (id) SETTINGS index_granularity = 8192;")
-		if err != nil {
-			return err
-		}
-
-		err = conn.Exec(context.Background(), "INSERT INTO replicated_test_table (id) VALUES (1);")
-		if err != nil {
-			return err
-		}
-
-		rows, err := conn.Query(context.Background(), "SELECT * FROM replicated_test_table;")
-		if err != nil {
-			return err
-		}
+func performReplicatedCRUD(t *testing.T, conn driver.Conn) ([]Test, error) {
+	return backoff.RetryNotifyWithData(
+		func() ([]Test, error) {
+			err := conn.Exec(context.Background(), "CREATE TABLE replicated_test_table (id UInt64) ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/mdb.data_transfer_cp_cdc', '{replica}') PRIMARY KEY (id) ORDER BY (id) SETTINGS index_granularity = 8192;")
+			if err != nil {
+				return nil, err
+			}
 
-		for rows.Next() {
-			var r Test
+			err = conn.Exec(context.Background(), "INSERT INTO replicated_test_table (id) VALUES (1);")
+			if err != nil {
+				return nil, err
+			}
 
-			err := rows.Scan(&r.Id)
+			rows, err := conn.Query(context.Background(), "SELECT * FROM replicated_test_table;")
 			if err != nil {
-				return err
+				return nil, err
 			}
 
-			res = append(res, r)
-		}
-		return nil
-	}, backoff.NewExponentialBackOff())
+			var res []Test
+			for rows.Next() {
+				var r Test
 
-	return res, err
-}
+				err := rows.Scan(&r.Id)
+				if err != nil {
+					return nil, err
+				}
 
-func performCRUD(conn driver.Conn) ([]Test, error) {
-	var (
-		err  error
-		rows []Test
+				res = append(res, r)
```

Comment:
Note that the previous version of this code is buggy - it reuses `res` between retries. This may lead to partial results from the failed attempt to be still in the slice.